File: src/main/java/model/core/guide/service/core/Member.java
package model.core.guide.service.core;

import model.core.guide.service.impl.CoreGuideFileServiceImpl;
import org.apache.commons.lang3.StringUtils;
import org.dom4j.Document;
import org.dom4j.Element;

import java.util.ArrayList;
import java.util.List;

public class Member extends Comm{
    public static final String PROJECT_ID = "memberTree";
    public static final String DEFAULT_ADMIN = "d456a11c-3028-46b6-8433-168b98b20c7a"; //默认管理员职务ID

    private static Document context = null;
    private static Element root = null;

    public Member(){

    }
    //清空缓存
    public void removeCache(){
        Member.context = null;
        Member.root = null;
    }

    //得到成员对象
    public Document getContext(){
        if(Member.context == null){
            Member.context = getXml(CoreGuideFileServiceImpl.getInstance().findOne(PROJECT_ID));
        }
        return Member.context;
    }
    //得到成员树根节点
    public Element getRoot(){
        if(Member.root == null){
            Member.root = getContext().getRootElement();
        }
        return Member.root;
    }
    //根据ID得到成员对象
    public Element getMemItem(String id){
        return getItem(getMemItems(id));
    }
    public List<Element> getMemItems(String id){
        if(StringUtils.isBlank(id)){
            return new ArrayList<>();
        }
        return getContext().selectNodes("//" + Comm.MEMBER_TAG + "[starts-with(@id,'"+ id +"')]");
    }

    //得到一个节点的所有兄弟节点
    public List<Element> getDescendantPerson(Element e){
        return e.selectNodes(".//" + Comm.MEMBER_TAG + "[@p='1']");
    }

    //根据成员ID判断是否为人
    public Boolean isPerson(String memberId){
        Boolean isPerson = false;
        List<Element> list = getMemItems(memberId);
        for(Element mem : list){
            String p = mem.attributeValue("p").toString();
            isPerson = StringUtils.isNotBlank(p) && Integer.parseInt(p) == 1;
            if(isPerson){
                break;
            }
        }
        return isPerson;
    }
    //从mem1中剥离出mem2
    public List<Element> stripping(Element e1,Element e2) {
        List<Element> memItems = new ArrayList<>();
        //将menu树中的mem装换成mem树中的mem
        e1 = getMemItem(getIdByItem(e1));
        if (contains(e1, e2,Comm.MEMBER_TAG)) {
            //得到mem1下一级节点
            List<Element> memChildren = getChildren(e1,Comm.MEMBER_TAG);
            for (Element oc : memChildren) {
                if (!equals(oc, e2)) {
                    List<Element> subOrgItems = stripping(oc, e2);
                    memItems.addAll(subOrgItems);
                }
            }
        }else {
            memItems.add(e1);
        }
        return memItems;
    }
}

File: src/main/webapp/model/core/guide/res/member.js
/**
 * Created by xc on 2018/2/7.
 */
let Mem = (function(){
    let dataUrl = "/core/guide/getXmlTree.do";

    let context;//xml对象
    let ParentsExpression;
    let ParentsAndSelfExpression;
    let SingleLevelChildrenExpression;
    let ChildrenExpression;
    let ChildrenAndSelfExpression;

    //固定的缓存对象
    let cache_root;
    let cache_mem_item = new Map();
    let cache_mem_parents = new Map();
    let cache_mem_parentsAndSelf = new Map();
    let cache_mem_singleLevelChildren = new Map();
    let cache_mem_children = new Map();
    let cache_mem_childrenAndSelf = new Map();

    //如果菜单上的成员发生改变，相应的要改变缓存中的值
    function removeCache(){
        cache_root = null;
        cache_mem_item = new Map();
        cache_mem_parents = new Map();
        cache_mem_parentsAndSelf = new Map();
        cache_mem_singleLevelChildren = new Map();
        cache_mem_children = new Map();
        cache_mem_childrenAndSelf = new Map();

        // console.log("-------------清空缓存-----------------");
    }

    //得到成员xml并转换成xsl
    function transform(url,param){
        if(!String.isNullOrWhiteSpace(url)){
            dataUrl = url
        }
        context = null;
        ParentsExpression = undefined;
        ParentsAndSelfExpression = undefined;
        SingleLevelChildrenExpression = undefined;
        ChildrenExpression = undefined;
        ChildrenAndSelfExpression = undefined;
        removeCache();
        return Comm.getXsltProcessor(Config.getMemTransformPath()).transformToFragment(getContext(param), document);
    }

    //得到成员xml
    function getContext(param) {
        if(context === undefined || context == null) {
            let p = {xmlType:"memberTree",projectId:"memberTree"};
            if(param){
                p = $.extend({}, p, param);
            }
            context = Comm.load(dataUrl, p);
        }
        return context;
    }

    //得到成员树根节点
    function getRoot() {
        if(cache_root == undefined || cache_root == null){
            cache_root = getContext().documentElement;
        }
        return cache_root;
    }

    //根据ID得到成员对象
    function getItem(id) {
        let val = cache_mem_item.get(id);
        if(Comm.isNull(val)){
            val = getContext().evaluate("//"+memberTag+"[@id='" + id + "']", getRoot(), null, XPathResult.FIRST_ORDERED_NODE_TYPE, null).singleNodeValue;
            cache_mem_item.set(id,val);
        }
        return val;
    }

    //得到父级节点
    function getParents(mem) {
        let id = Comm.getIdByItem(mem);
        let val = cache_mem_parents.get(id);
        if(Comm.isNull(val)){
            val = Comm.getParents(getContext(), mem, ParentsExpression);
            cache_mem_parents.set(id,val);
        }
        return val;
    }

    //得到父级节点以及自己
    function getParentsAndSelf(mem) {
        let id = Comm.getIdByItem(mem);
        let val = cache_mem_parentsAndSelf.get(id);
        if(Comm.isNull(val)){
            val = Comm.getParentsAndSelf(getContext(), mem, ParentsAndSelfExpression);
            cache_mem_parentsAndSelf.set(id,val);
        }
        return val;
    }

    //得到该节点的下一级子节点
    function getSingleLevelChildren(mem) {
        let id = Comm.getIdByItem(mem);
        let val = cache_mem_singleLevelChildren.get(id);
        if(Comm.isNull(val)){
            if (SingleLevelChildrenExpression === undefined) {
                SingleLevelChildrenExpression = getContext().createExpression("child::"+memberTag, null);
            }
            val = SingleLevelChildrenExpression.evaluate(mem, XPathResult.ORDERED_NODE_SNAPSHOT_TYPE, null);
            cache_mem_singleLevelChildren.set(id,val);
        }
        return val;
    }

    //得到所有子节点
    function getChildren(mem) {
        let id = Comm.getIdByItem(mem);
        let val = cache_mem_children.get(id);
        if(Comm.isNull(val)){
            if (ChildrenExpression === undefined) {
                ChildrenExpression = getContext().createExpression("descendant::"+memberTag, null);
            }
            val = ChildrenExpression.evaluate(mem, XPathResult.ORDERED_NODE_SNAPSHOT_TYPE, null);
            cache_mem_children.set(id,val);
        }
        return val;
    }

    //得到所有子节点以及自己
    function getChildrenAndSelf(mem) {
        let id = Comm.getIdByItem(mem);
        let val = cache_mem_childrenAndSelf.get(id);
        if(Comm.isNull(val)){
            if (ChildrenAndSelfExpression === undefined) {
                ChildrenAndSelfExpression = getContext().createExpression("descendant-or-self::"+memberTag, null);
            }
            val = ChildrenAndSelfExpression.evaluate(mem, XPathResult.ORDERED_NODE_SNAPSHOT_TYPE, null);
            cache_mem_childrenAndSelf.set(id,val);
        }
        return val;
    }

    //根据节点id，判断一个节点是否与另一个节点相等
    function equals(mem1, mem2) {
        return Comm.equals(mem1, mem2);
    }

    //判断一个mem1节点是否包含另一个mem2节点
    function contains(mem1, mem2) {
        let result = getContext().evaluate(".//"+memberTag+"[@id='" + Comm.getIdByItem(mem2) + "']", mem1, null, XPathResult.ORDERED_NODE_SNAPSHOT_TYPE, null);
        return result != null && Comm.getLength(result) > 0 ? true : false;
    }

    //从mem1中剥离出mem2
    function stripping(mem1, mem2) {
        let memItems = new Array();
        //将menu树中的mem装换成mem树中的mem
        mem1 = getItem(Comm.getIdByItem(mem1))
        if (contains(mem1, mem2)) {
            //得到mem1下一级节点
            let memChildren = getSingleLevelChildren(mem1);
            for (let i = 0; i < Comm.getLength(memChildren); i++) {
                let oc = Comm.getItem(memChildren, i);
                //两个
                if (!equals(oc, mem2)) {
                    let subOrgItems = stripping(oc, mem2);
                    memItems = memItems.concat(subOrgItems);
                }
            }
        }else {
            memItems.push(mem1);
        }
        return memItems;
    }

    return {
        removeCache : removeCache,
        transform : transform,
        getContext : getContext,
        getRoot : getRoot,
        getItem : getItem,
        //getParent : getParent,
        getParents : getParents,
        getParentsAndSelf : getParentsAndSelf,
        getSingleLevelChildren : getSingleLevelChildren,
        getChildren : getChildren,
        getChildrenAndSelf : getChildrenAndSelf,
        stripping : stripping,
        equals : equals,
        contains : contains
    }
})();